src/app/access-requests/features/access-request-field-edit/access-request-field-edit.component.ts

```diff
@@ -38,58 +47,249 @@ import { ConfigService } from '@app/shared/services/config.service';
-    this.name() == 'ticket_id' && this.field()
-      ? this.#baseTicketUrl + this.field()
-      : null,
+  ticketUrl = computed<(string | null)[]>(() =>
+    this.names().map((x, idx) =>
+      x == 'ticket_id' && this.fields()[idx]
+        ? this.#baseTicketUrl + this.fields()[idx]
+        : null,
+    ),
   );
 
-  #defaultValue = computed<string>(() => (this.request() as any)[this.name()] || '');
+  todayMidnight = new Date();
+  minFromDate = new Date();
+  minUntilDate = new Date();
+  maxFromDate = new Date();
+  maxUntilDate = new Date();
+  fieldErrorMessages = signal<string[]>([]);
+
+  #defaultValues = computed<string[]>(() =>
+    this.names().map((x) => this.request()[x] || ''),
+  );
+
+  durationFormControl: (FormControl | null)[] = [];
+
+  constructor() {
+    this.todayMidnight.setHours(0, 0, 0, 0);
+    let d = new Date();
+    d.setDate(d.getDate() + this.#config.defaultAccessDurationDays);
+    this.minUntilDate.setDate(
+      this.minUntilDate.getDate() + this.#config.accessGrantMinDays,
+    );
+    this.updateAccessEndRanges(new Date());
+    this.updateAccessStartRanges(d);
+  }
+
+  /**
+   * Populate the editable fields with the values from the access request on component init.
+   */
+  ngOnInit(): void {
+    this.fields.update(() =>
+      this.#defaultValues().map((x, idx) => {
+        if (this.names()[idx] === 'access_starts' && new Date(x) < new Date())
+          return new Date();
+        return x;
+      }),
+    );
+
+    this.durationFormControl = this.names().map((x, idx) => {
+      if (x === 'access_starts')
+        return new FormControl('', [
+          Validators.required,
+          (control) => this.#fromDateValidator(control, idx),
+        ]);
+      else if (x === 'access_ends')
+        return new FormControl('', [
+          Validators.required,
+          (control) => this.#untilDateValidator(control, idx),
+        ]);
+      return null;
+    });
+    this.fieldErrorMessages.set(Array(this.names().length).fill(''));
+  }
+
+  /**
+   * Update the range for the until date based on the from date
+   * @param date - The from date to update the range for
+   */
+  updateAccessEndRanges(date: Date): void {
+    const newUntilMinDate = new Date(
+      date.getTime() + this.#config.accessGrantMinDays * MILLISECONDS_PER_DAY,
+    );
+    const newUntilMaxDate = new Date(
+      date.getTime() + this.#config.accessGrantMaxDays * MILLISECONDS_PER_DAY,
+    );
+
+    this.minUntilDate = newUntilMinDate;
+    this.maxUntilDate = newUntilMaxDate;
+  }
+
+  /**
+   * Update the range for the from date based on the until date
+   * @param date - The until date to update the range for
+   */
+  updateAccessStartRanges(date: Date): void {
+    const currentDate = this.todayMidnight;
+
+    const newFromMinDate = new Date(
+      Math.max(
+        currentDate.getTime(),
+        date.getTime() - this.#config.accessGrantMaxDays * MILLISECONDS_PER_DAY,
+      ),
```

Comment:
Here you're applying the (stricter) restrictions for the _users_ when doing access requests. The _data stewards_ have less strict restrictions:

- start can be today or later, `accessUpfrontMaxDays` is not enforced in this case
- duration must be `accessGrantMaxDays * accessGrantMaxExtend` days at max

See our implementation in the access request service. Note the additional factor `accessGrantMaxExtend` that allows the data stewards to grant longer time periods than users can request if needed. This still needs to be added as config setting to the frontend. Currently the setting is only used in the access request service.